### PR TITLE
Cr 1496 client ip updates

### PR DESCRIPTION
**Pull request by Paul-Joel**
# Motivation and Context
Change to generation of clientIP for rate limiting purposes

# What has changed
clientIP is now passed through a regex to check it only contains ip like content, and to take value third from right instead of the first value.
Added a unit test to check the validity tests

No Cucumber updates required

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1496

# Screenshots (if appropriate):

### CODE REVIEW

File: app/utils.py [hunk 1]
```diff
@@ -52,7 +52,14 @@ def get_now_utc():
     @staticmethod
     def single_client_ip(request):
         if request['client_ip']:
-            single_ip = request['client_ip'].split(',', 1)[0]
+            client_ip = request['client_ip']
+            for character in string.whitespace + OBSCURE_WHITESPACE:
+                client_ip = client_ip.replace(character, '')
+            ip_validation_pattern = re.compile(r'^[0-9.,\s]*$')
+            if ip_validation_pattern.fullmatch(client_ip):
```
Comment (rpcatling) [inline app/utils.py:59]: consider this enhancement:
```
 if ip_validation_pattern.fullmatch(client_ip) and client_ip.count(',') > 1:
```
just in case the expected format coming is in different

File: app/utils.py [hunk 2]
```diff
@@ -52,7 +52,14 @@ def get_now_utc():
...
+                single_ip = client_ip.split(',', -1)[-3]
+            else:
```
Comment (rpcatling) [inline app/utils.py:61]: should we log a warning here ?

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -329,3 +329,16 @@ def test_check_before_after_census_day(self):
         self.after_census_day(2021, 3, 22)
         self.before_census_day(2021, 3, 21)
         self.before_census_day(2021, 3, 20)
+
+    def test_client_ip(self):
```
Comment (rpcatling) [inline tests/unit/test_utils.py:333]: Consider enhancing for:
- inputs with and without various whitespace
- inputs with no client_ip in the request
- inputs with just one client_ip in the request

**Review by rpcatling — CHANGES_REQUESTED** [review]
some improvements requested

File: tests/unit/test_utils.py [hunk 4]
```diff
@@ -329,3 +329,16 @@ def test_check_before_after_census_day(self):
...
+        valid_request = {'client_ip': '192.168.0.0, 35.190.0.0, 35.191.10.0'}
+        spoofed_request = {'client_ip': '200.30.20.10, 192.168.0.0, 35.190.0.0, 35.191.10.0'}
+        invalid_request = {'client_ip': 'GHT.RGT.ED.WS, 192.168.0.0, 35.190.0.0, 35.191.10.0'}
+        built_valid = View.single_client_ip(valid_request)
+        built_invalid = View.single_client_ip(invalid_request)
+        built_spoofed = View.single_client_ip(spoofed_request)
+        expected_valid = '192.168.0.0'
+        expected_empty = ''
+        self.assertEqual(built_valid, expected_valid)
+        self.assertEqual(built_spoofed, expected_valid)
```
Comment (cookel2) [inline tests/unit/test_utils.py:343]: Should a spoofed request come through as expected_valid? I'm not sure I understand this.

**Review by cookel2 — APPROVED** [review]
It looks good except that I don't understand the spoofed_request test. Is it just proving that it will ignore the first ip (unless it's the third one from the end)?

File: app/utils.py (hunk 1 above)
Comment (Paul-Joel) [reply app/utils.py]: Added this in

File: app/utils.py (hunk 2 above)
Comment (Paul-Joel) [reply app/utils.py:59]: Warning added

File: tests/unit/test_utils.py (hunk 3 above)
Comment (Paul-Joel) [reply tests/unit/test_utils.py:333]: I've added a couple more tests, and got rid of the whitespace code (I realised I was doing the split incorrectly)

**Review by rpcatling — APPROVED** [review]
LGTM
